### PR TITLE
Verify nodes' str() and repr() don't raise errors/warnings

### LINKED ISSUE #1881: Some `__str__` methods of nodes raise errors and warnings
### Steps to reproduce
```python
>>> fd = astroid.nodes.FunctionDef()
>>> str(fd)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jacobwalls/astroid/astroid/nodes/node_ng.py", line 219, in __str__
    value = getattr(self, field)
            ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'FunctionDef' object has no attribute 'args'
```

Less severely, but still not great, launching python with warnings e.g. `python -Wall`:
```python
>>> cd = astroid.nodes.ClassDef()
>>> str(cd)
/Users/.../astroid/astroid/nodes/scoped_nodes/scoped_nodes.py:2041: DeprecationWarning: The 'ClassDef.doc' attribute is deprecated, use 'ClassDef.doc_node' instead.
```

***
We should add a unittest that generates the `str()` and `repr()` of all node types.

### `python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"` output
2.12.13

### PR DESCRIPTION
**Pull request by crazybolillo**
Calling str() or repr() on certain nodes fails either with errors or warnings. This commit adds unittests to find and reproduce this behavior with the endgoal of fixing it.

Closes #1881

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

### CODE REVIEW

**Comment by crazybolillo** [comment]
I changed the test flavor to pytest and fixed all `str` and `repr` issues found in the tests. After all I had to add some more failsafes in other places.

**Comment by crazybolillo** [comment]
Also something weird is going on with the PR, since it is trying to reapply a commit already found in the main branch (same hash and everything), I guess it has to do with me not having some sort of permission to update the PR or execute a certain flow.

**Comment by Pierre-Sassoulas** [comment]
Could you rebase or merge main in your branch ?

**Comment by crazybolillo** [comment]
> Could you rebase or merge main in your branch ?

It should be. My fork shows my branch being just one commit ahead and none behind. :open_mouth:

File: tests/test_nodes.py
Comment (DanielNoord) [inline tests/test_nodes.py:1906]: I like this trick, never thought of doing it like this!

File: astroid/nodes/scoped_nodes/scoped_nodes.py
Comment (DanielNoord) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:1353]: I'd prefer just setting `decorators` in the ``__init__``. Since it can be `None` anyway this doesn't interfere with the typing of the attribute.

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [inline astroid/nodes/node_ng.py:249]: For which of the nodes is this necessary? `get_children` isn't only used in `__str__` so I'm worried people might depend on this returning an `AttributeError`.

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [inline astroid/nodes/node_ng.py:213]: :+1:

File: astroid/nodes/node_classes.py
Comment (DanielNoord) [inline astroid/nodes/node_classes.py:861]: See other comment. If we could find a way where we only impact `__str__` and `__repr__` I would be very much in favour of that.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks for doing this! I really like the test you came up with!

**Comment by crazybolillo** [comment]
Ah I see, makes sense to only touch `__str__` and `__repr__`.  Apparently the only node that had issues after patching `value = getattr(self, field)` was `MatchCase`.

The error is the following:
`AttributeError: 'MatchCase' object has no attribute 'pattern'`

I solved inside `__repr__` so nothing else is modified now.

Thanks for the review :handshake:

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [inline astroid/nodes/node_ng.py:232]: Shall we add a comment like:
`# If the __postinit__ isn't called `fromlineno` could not exist`.

Although I must say, shouldn't `fromlineno` as property always exist?

Comment (crazybolillo) [reply astroid/nodes/node_ng.py:232]: As far as my understanding goes, `fromlineno` does always exist. The problem is that it calls `fixed_source_line` and thats where the `AttributeError` is generated, specifically it seems to come from `get_children`.

I ran this on a different dev machine and got the error for a few other nodes besides `MatchCase` (weird). Here are some sample stack traces:

```
self = <[AttributeError("'Arguments' object has no attribute 'posonlyargs'") raised in repr()] Arguments object at 0x7f4c48347f90>

    def get_children(self):
>       yield from self.posonlyargs or ()
E       AttributeError: 'Arguments' object has no attribute 'posonlyargs'
```

```
@cached_property
def fromlineno(self) -> int:
....
>       if self.decorators is not None:
E       AttributeError: 'AsyncFunctionDef' object has no attribute 'decorators'
```

I guess the comment could be something like `The dependencies used to calculate fromlineno (if not cached) may not exist at the time`?

Comment (crazybolillo) [reply astroid/nodes/node_ng.py:232]: Adding to my previous comment, in more general terms it seems `fromlineno` may execute certain function calls in an attempt to obtain the line number, and the resources used in said functions may not exist at the time of the read attempt.

Comment (DanielNoord) [reply astroid/nodes/node_ng.py:232]: Yeah you suggestion for the comment seems fine! Thanks!

File: tests/test_nodes.py
Comment (jacobtylerwalls) [inline tests/test_nodes.py:1895]: Won't block this merge over it, but stylistically I'd suggest a comprehension instead of listifying a filter.

```suggestion
    [
        node
        for node in astroid.nodes.ALL_NODE_CLASSES
        if node.__name__
        not in ["_BaseContainer", "BaseContainer", "NodeNG", "const_factory"]
    ],
```

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for taking this on! Will approve and merge once the comment is added that's discussed above.

**Review by DanielNoord — APPROVED** [review]
Thanks for picking this up!

**Pull request merged**
